### PR TITLE
Switch to serde, and other major changes

### PR DESCRIPTION
**Pull request by erickt**
This is a breaking change. This switches mustache over to using serde, renames some variants, and stops publicly exposing some of the interior modules. It also extracts out many of the tests.

### CODE REVIEW

File: tests/encoder.rs [hunk 1]
```diff
@@ -1,31 +0,0 @@
...
-
-#[test]
-fn encodable_with_no_data() {
-    assert_let!(Err(EncoderError::NoDataToEncode) = encode(&NoData));
-}
-
-#[test]
-fn encodable_with_multiple_roots() {
-    assert_let!(Err(EncoderError::MultipleRootsFound) = encode(&Multiroot));
-}
```
Comment (Ryman) [inline tests/encoder.rs]: Are there no equivalent tests for these, or are they impossible to construct in serde 1.0?

File: src/lib.rs [hunk 2]
```diff
@@ -53,92 +34,6 @@ where
...
-            Data::Map(ref v) => write!(f, "Map({:?})", v),
-            Data::Fun(_) => write!(f, "Fun(...)"),
-        }
-    }
-}
-
-/// Represents the shared metadata needed to compile and render a mustache
-/// template.
-#[derive(Clone)]
-pub struct Context {
```
Comment (Ryman) [inline src/lib.rs]: Looks like this and the impls for Data are missing from the commit? CI suggests a missing `context.rs` file.

File: src/encoder.rs [hunk 3]
```diff
@@ -136,11 +136,11 @@ impl serde::Serializer for Encoder {
...
         _name: &'static str,
         _variant_index: u32,
-        _variant: &'static str,
+        variant: &'static str,
     ) -> Result<Data>
     {
         // FIXME: Perhaps this could be relaxed to just 'do nothing'
-        Err(Error::UnsupportedType)
+        Ok(Data::StrVal(variant.to_string()))
     }
```
Comment (Ryman) [inline src/encoder.rs]: nit: drop comment and squash commit

**Comment by cardoe** [comment]
This fails to build with:
```
error[E0583]: file not found for module `context`
  --> src/lib.rs:13:5
    |
13  | mod context;
```

**Comment by Ryman** [comment]
Merged a rebase of this in https://github.com/nickel-org/rust-mustache/pull/55, thanks for the work here!

**Pull request closed without merging**